model/generate.go

```diff
@@ -102,51 +103,70 @@ func ParseProjectFromJSON(data []byte) (GeneratedProject, error) {
 	}
 	if v == nil {
-		return nil, nil, nil, nil, gimlet.ErrorResponse{StatusCode: http.StatusBadRequest, Message: errors.Wrapf(err, "unable to find version %s", t.Version).Error()}
+		return nil, nil, nil, nil, "",
+			gimlet.ErrorResponse{StatusCode: http.StatusBadRequest, Message: errors.Wrapf(err, "unable to find version %s", t.Version).Error()}
 	}
+	prevConfig := v.Config
 	p := &Project{}
 	if err = LoadProjectInto([]byte(v.Config), t.Project, p); err != nil {
-		return nil, nil, nil, nil, gimlet.ErrorResponse{StatusCode: http.StatusBadRequest, Message: errors.Wrap(err, "error reading project yaml").Error()}
+		return nil, nil, nil, nil, "",
+			gimlet.ErrorResponse{StatusCode: http.StatusBadRequest, Message: errors.Wrap(err, "error reading project yaml").Error()}
 	}
 
 	// Cache project data in maps for quick lookup
 	cachedProject := cacheProjectData(p)
 
 	// Validate generated project against original project.
 	if err = g.validateGeneratedProject(p, cachedProject); err != nil {
-		return nil, nil, nil, nil, gimlet.ErrorResponse{StatusCode: http.StatusBadRequest, Message: errors.Wrap(err, "generated project is invalid").Error()}
+		return nil, nil, nil, nil, "",
+			gimlet.ErrorResponse{StatusCode: http.StatusBadRequest, Message: errors.Wrap(err, "generated project is invalid").Error()}
 	}
 
 	config, err := g.addGeneratedProjectToConfig(v.Config, cachedProject)
 	if err != nil {
-		return nil, nil, nil, nil, gimlet.ErrorResponse{StatusCode: http.StatusBadRequest, Message: errors.Wrap(err, "error creating config from generated config").Error()}
+		return nil, nil, nil, nil, "",
+			gimlet.ErrorResponse{StatusCode: http.StatusBadRequest, Message: errors.Wrap(err, "error creating config from generated config").Error()}
 	}
 	v.Config = config
 	if err := LoadProjectInto([]byte(v.Config), t.Project, p); err != nil {
-		return nil, nil, nil, nil, gimlet.ErrorResponse{StatusCode: http.StatusBadRequest, Message: errors.Wrap(err, "error reading project yaml").Error()}
+		return nil, nil, nil, nil, "",
+			gimlet.ErrorResponse{StatusCode: http.StatusBadRequest, Message: errors.Wrap(err, "error reading project yaml").Error()}
 	}
-	return p, v, t, &cachedProject, nil
+	return p, v, t, &cachedProject, prevConfig, nil
 }
 
-func (g *GeneratedProject) Save(p *Project, v *version.Version, t *task.Task, pm *projectMaps) error {
-	if err := version.UpdateOne(bson.M{version.IdKey: v.Id}, bson.M{"$set": bson.M{version.ConfigKey: v.Config}}); err != nil {
+func (g *GeneratedProject) Save(p *Project, v *version.Version, t *task.Task, pm *projectMaps, prevConfig string) error {
+	query := bson.M{version.IdKey: v.Id}
+	if prevConfig != "" {
+		query[version.ConfigKey] = prevConfig
+	}
+	update := bson.M{"$set": bson.M{version.ConfigKey: v.Config}}
+
+	err := version.UpdateOne(query, update)
+	if err == mgo.ErrNotFound {
```

Comment:
I'm concerned that someone may come along and wrap this, since it's not obvious from here why this branch exists. It might be clearer to use `errors.Cause` to unwrap the error in the caller and get rid of this if/else.